Add Nacos Support

KonaEspresso94 opened the pull request:
Motivation:

Nacos is a service discovery application extensively used. By providing native support for Nacos, it will be beneficial to Armeria users.

#5360 #5365


Modification:

- Add a `nacos` module.
- Add `NacosClient`. 
Its internal implementation includes `LoginClient` for obtaining an accessToken through basic Nacos authentication, `QueryInstancesClient` for querying service instances, and `RegisterInstanceClient` for handling the registration and deregistration of instances.
- Implement `NacosEndpointGroup` and `NacosUpdatingListener`.
- Configure unit tests using Testcontainers and the Nacos image to facilitate actual call testing.


Result:

Closes #5365

Nacos discovery and instance registration functionalities are now integrated into Armeria.

---

In writing this code, I was strongly influenced by the implementation on the consul and eureka modules. 
I appreciate your review and feedback. Thank you.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=5409) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=5409) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=5409) it.</sub>

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/LoginClient.java:
```diff
@@ -0,0 +1,130 @@
+ * A Nacos client that is responsible for
+ * <a href="https://nacos.io/en-us/docs/auth.html">Nacos Authentication</a>.
+ */
+final class LoginClient {
```
The decorator pattern may hide the details of authentication.
How about making `LoginClient` as a decorator to decorate `WebClient`?

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/LoginClient.java:
```diff
@@ -0,0 +1,130 @@
+
+    private static final String NACOS_ACCESS_TOKEN_CACHE_KEY = "NACOS_ACCESS_TOKEN_CACHE_KEY";
+    private final AsyncLoadingCache<String, LoginResult> tokenCache = Caffeine.newBuilder()
+            .maximumSize(1)
```
Optional) It seems overkill to use Caffeine only for 1 element.
What do you think of creating a function that triggers `loginInternal()` if access tokens are expired while invoking `login()`?

If you find it more difficult than Caffeine approach, I'm okay to keep this as is.

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/QueryInstancesClient.java:
```diff
@@ -0,0 +1,223 @@
+        final StringBuilder path = new StringBuilder("/").append(nacosApiVersion).append("/ns/instance/list?");
+        final QueryParamsBuilder paramsBuilder = NacosClientUtil
+                .queryParamsBuilder(namespaceId, groupName, serviceName, clusterName, healthyOnly, app, null,
+                                    null, null);
```
It would be inefficient to create `QueryParams` for the same parameters for all requests.
How about creating and caching the `QueryParams` as a member field when `QueryInstancesClient` is created?

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/QueryInstancesClient.java:
```diff
@@ -0,0 +1,223 @@
+                    .as(HttpEntity::content)
+                    .execute();
+        } else {
+            return loginClient.login().thenCompose(accessToken ->
```
As I mentioned above, would you make `loginClient` as a decorator?

ikhoon reviewed (commented):
The overall direction looks great.

KonaEspresso94 commented:
@ikhoon 

Thank you for your review comments. I've applied your suggestions into the code. 

- `LoginClient` now implements `SimpleDecoratingHttpClient`. Contexts related to 'nacos basic authentication' have been hidden as much as possible from other clients.
- Caching access token by own implementation instead of Caffeine cache. 
- Changed the query string that was being created each time in `LoginClient` and `QueryInstanceClient` to be reused efficiently. 
- `NacosUpdatingListener` and `RegisterInstanceClient` are not guaranteed to always have the same endpoint used as parameter. Eventually, we will need to create a `StringBuilder` or `QueryParamsBuilder`. However, I have minimized the parameter of the API by turning immutable parameters into properties, such as in the above two clients. 
- Fixed some missing `final`.


The implementation to replace the Caffein, I made it as simple as possible. So now the login api can be called concurrently. Considering the usability of `NacosUpdatingListener` and `NacosEndpointGroup`, I think it's unlikely that concurrent calls will actually happen, and if they do, it won't be an issue. But I'm not sure. I'd be interested in your thoughts. 

Thank you!

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/LoginClient.java:
```diff
@@ -122,6 +91,39 @@ public HttpResponse execute(ClientRequestContext ctx, HttpRequest req) {
+                return UnmodifiableFuture.completedFuture(cacheEntry.loginResult.accessToken);
+            }
+
+            return login().thenApply(loginResult -> {
```
This may result in sending multiple login requests that are not what we really want.
If `login()` is being performed, other requests can piggyback on the returned future.

By the way, I think an `AsyncLoader` loading a value when it is expired or periodically would be useful in many places.
Created https://github.com/line/armeria/issues/5506

trustin replied:
Maybe we shouldn't let @KonaEspresso94 wait for it to be merged too long and use Caffeine internally and later replace the Caffeine usage with `AsyncLoader`?

ikhoon replied:
I didn't know it would take this long. 😅 Let's leave that part as is. I will continue to review this PR tomorrow.

KonaEspresso94 replied:
Thank you for your consideration. Following your comments, I have reverted the commit to use Caffeine instead of my own implementation.

trustin commented:
@KonaEspresso94 Gentle ping - looking forward to having this feature merged :bow:

jrhee17 commented on nacos/src/main/java/com/linecorp/armeria/common/nacos/NacosConfigSetters.java:
```diff
@@ -0,0 +1,43 @@
+ * Sets properties for building a Nacos client.
+ */
+@UnstableApi
+public interface NacosConfigSetters {
```
optional; What do you think of generalizing this interface and adding the related methods?

e.g.
```java
public interface NacosConfigSetters<SELF extends NacosConfigSetters<SELF>> {
  SELF nacosApiVersion(...)
}
```

jrhee17 commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/QueryInstancesClient.java:
```diff
@@ -0,0 +1,211 @@
+    CompletableFuture<List<Endpoint>> endpoints() {
+        return queryInstances()
+                .thenApply(response -> response.data.hosts.stream()
+                        .map(QueryInstancesClient::toEndpoint)
+                        .filter(Objects::nonNull)
+                        .collect(toImmutableList()));
```
Do you mind running an auto-indent for files in this PR in general one more time?

jrhee17 commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/QueryInstancesClient.java:
```diff
@@ -0,0 +1,211 @@
+
+    CompletableFuture<List<Endpoint>> endpoints() {
+        return queryInstances()
+                .thenApply(response -> response.data.hosts.stream()
```
Question) Just to make sure, even if `code != 0` (which I assume means successful) will `data` and `hosts` always exist?
I'm asking because there doesn't seem to defence against `null` check

KonaEspresso94 replied:
@jrhee17 

I added null check for data and data.hosts! Thanks.

jrhee17 commented on nacos/src/test/java/com/linecorp/armeria/internal/nacos/NacosTestBase.java:
```diff
@@ -0,0 +1,152 @@
+        return nacosUri;
+    }
+
+    protected static int[] unusedPorts(int numPorts) {
```
To other reviewers) I guess the ideal approach (binding to port 0) can be done for all similar tests at once. I think this code is fine if the CI passes reliably at least for now.

jrhee17 commented on nacos/src/main/java/com/linecorp/armeria/client/nacos/NacosEndpointGroupBuilder.java:
```diff
@@ -0,0 +1,183 @@
+                nacosClientBuilder.build(), registryFetchIntervalMillis);
+    }
+
+    @Override
+    public NacosEndpointGroupBuilder allowEmptyEndpoints(boolean allowEmptyEndpoints) {
+        return (NacosEndpointGroupBuilder) super.allowEmptyEndpoints(allowEmptyEndpoints);
+    }
+
+    /**
+     * Sets the timeout to wait until a successful {@link Endpoint} selection.
+     * {@link Duration#ZERO} disables the timeout.
+     * If unspecified, {@link Flags#defaultResponseTimeoutMillis()} is used by default.
+     */
+    @Override
+    public NacosEndpointGroupBuilder selectionTimeout(Duration selectionTimeout) {
+        return (NacosEndpointGroupBuilder) super.selectionTimeout(selectionTimeout);
+    }
+
+    /**
+     * Sets the timeout to wait until a successful {@link Endpoint} selection.
+     * {@code 0} disables the timeout.
+     * If unspecified, {@link Flags#defaultResponseTimeoutMillis()} is used by default.
+     */
+    @Override
+    public NacosEndpointGroupBuilder selectionTimeoutMillis(long selectionTimeoutMillis) {
+        return (NacosEndpointGroupBuilder) super.selectionTimeoutMillis(selectionTimeoutMillis);
+    }
```
This is unrelated to your PR, but we can probably remove these lines now

KonaEspresso94 replied:
@jrhee17 
I have removed the code from `NacosEndpointGroupBuilder`. Should I also apply this to other extended classes of `AbstractDynamicEndpointGroupBuilder` on this PR?

jrhee17 reviewed (commented):
Looks good overall 👍 Left some minor comments

jrhee17 reviewed (approved):
Only nits for me, thanks @KonaEspresso94 👍 👍 👍

minwoox reviewed (commented):
Looks pretty great! 👍

minwoox reviewed (commented):
Looks pretty great! 👍

KonaEspresso94 commented on nacos/src/main/java/com/linecorp/armeria/client/nacos/NacosEndpointGroupBuilder.java:
```diff
@@ -0,0 +1,139 @@
+        extends AbstractDynamicEndpointGroupBuilder<NacosEndpointGroupBuilder>
+        implements NacosConfigSetters<NacosEndpointGroupBuilder> {
+
+    private static final long DEFAULT_CHECK_INTERVAL_MILLIS = 30_000;
+    private final NacosClientBuilder nacosClientBuilder;
```
@minwoox 

Oops. Why did I set this to 30secs? 
I'll change it to 10secs like the ConsulEndpointGroup. thx.

KonaEspresso94 commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/NacosClientBuilder.java:
```diff
@@ -0,0 +1,126 @@
+    }
+
+    public NacosClient build() {
+        final URI uri;
+        try {
+            uri = new URI(nacosUri.getScheme(), null, nacosUri.getHost(), nacosUri.getPort(),
+                          "/nacos", null, null);
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e);
+        }
```
@jrhee17 

You're suggesting that `uri = nacosUri.resolve("/nacos")` would suffice, right? 
I wasn't aware of this usecase of URI class. 
I'll make the change. Thx.

KonaEspresso94 commented on nacos/src/main/java/com/linecorp/armeria/client/nacos/NacosEndpointGroup.java:
```diff
@@ -0,0 +1,134 @@
+    private final EventExecutor eventLoop;
+
+    @Nullable
+    private ScheduledFuture<?> scheduledFuture;
```
@minwoox @jrhee17 

Hmm. Actually, I initially used volatile, but changed to the current implementation based on [other review comment](https://github.com/line/armeria/pull/5409#discussion_r1642097222).
Considering low call rate of this logic path, there shouldn't be a significant performance difference either way. 
It comes down to choosing whether to depend on EventLoop for more readable code or remove it and use slightly more verbose code.

The commit : https://github.com/line/armeria/pull/5409/commits/7476edec045c6d78925d0a32f82317f9301c367b

minwoox replied:
Ah, then let's keep the current implementation. 😉

KonaEspresso94 commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/NacosClient.java:
```diff
@@ -0,0 +1,105 @@
+
+        if (username != null && password != null) {
+            builder.decorator(LoginClient.newDecorator(builder.build(), username, password));
+        }
```
@jrhee17 

Oh, that wasn't intended. 
Also the builder wasn't properly passed along in the builder-chain. 
I'll make sure to refine it overall. Thank you!

KonaEspresso94 commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/NacosClientBuilder.java:
```diff
@@ -0,0 +1,126 @@
+        final URI uri;
+        try {
+            uri = new URI(nacosUri.getScheme(), null, nacosUri.getHost(), nacosUri.getPort(),
+                          "/nacos", null, null);
```
@minwoox 

While /nacos is the default path, but it could be behind a proxy and there is no guarantee that it cannot be changed through configuration. Therefore, it is better to allow users to configure it directly. 

I'll remove it. Thanks.

KonaEspresso94 commented on nacos/src/main/java/com/linecorp/armeria/common/nacos/NacosConfigSetters.java:
```diff
@@ -0,0 +1,75 @@
+     * @param username the username for access Nacos API, default: {@code null}
+     * @param password the password for access Nacos API, default: {@code null}
+     */
+    SELF authorization(String username, String password);
```
@minwoox 

According to the [Nacos documentation](https://nacos.io/en-us/docs/v2/plugin/auth-plugin.html), there is no official support for generating access tokens directly. However, it can be added through a customized auth plugin. Since the specific requirements of the custom plugin are unpredictable, it seems preferable to allow users to customize decorator instead of provided LoginClient decorator.

Also, would it be beneficial to provide an implementation for Aliyun(Alibaba Cloud spec?) Auth?

https://nacos.io/en-us/docs/v2/plugin/auth-plugin.html

minwoox replied:
I didn't know that requires additional plugin. 😓 
Let's keep as is and implement the plugin later if there's a demand for that. 😉

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/server/nacos/NacosUpdatingListenerBuilder.java:
```diff
@@ -0,0 +1,108 @@
+        nacosClientBuilder.authorization(username, password);
+        return this;
+    }
+
```
There is `useHealthyEndpoints()` method on the client side.
Should we also support `health` status when registering an instance?
https://nacos.io/en-us/docs/open-api.html#2.1

ikhoon commented on nacos/src/main/java/com/linecorp/armeria/internal/nacos/LoginClient.java:
```diff
@@ -0,0 +1,159 @@
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                              .omitNullValues()
+                              .add("accessToken", accessToken)
```
Changed to use `AuthToken` because sensitive information should not be excluded from serialization.

ikhoon reviewed (approved):
Thanks, @KonaEspresso94. 🚀🙇‍♂️
Left a minor suggestion.

jrhee17 commented on nacos/src/main/java/com/linecorp/armeria/server/nacos/NacosUpdatingListener.java:
```diff
@@ -0,0 +1,144 @@
+ * <a href="https://nacos.io">Nacos</a>.
+ */
+@UnstableApi
+public class NacosUpdatingListener extends ServerListenerAdapter {
```
```suggestion
public final class NacosUpdatingListener extends ServerListenerAdapter {
```

minwoox reviewed (approved):
@KonaEspresso94 Sorry about the late review. Thanks a lot! 👍 👍 👍
